lib/core/src/Cardano/Wallet/Address/Pool.hs

```diff
@@ -238,3 +247,24 @@ ensureFresh ix pool@Pool{generator,gap,addresses}
         to = toEnum $ fromEnum ix + fromIntegral gap - 1
         -- example:
         --  ix = 0 && fresh = 0 && gap = 20 `implies` [fresh .. to] = [0..19]
+
+{-------------------------------------------------------------------------------
+    Address discovery using an address pool
+-------------------------------------------------------------------------------}
+-- | Discover transactions and addresses by
+-- using an efficient query @addr -> m txs@ and an address pool.
+discover
+    :: (Enum ix, Ord addr, Monad m, Monoid txs, Eq txs)
+    => (addr -> m txs) -> Pool addr ix -> m (txs, Pool addr ix)
+discover query pool0 =
+    go mempty pool0 $ toEnum 0
+  where
+    go !txs1 !pool1 old = do
+        -- TODO: Maybe cache the `generator` in the Pool using lazy evaluation.
+        let addr = generator pool0 old
```

Comment:
Generator is a bit generic, indeed. The name `fromIx` feels a bit like it would be a simple conversion, but it actually involves "expensive" cryptography operations, hence I feel that this is too short. Perhaps `addressFromIx`? 🤔